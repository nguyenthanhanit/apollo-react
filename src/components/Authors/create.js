import React, {useState} from 'react';
import {gql, useMutation} from '@apollo/client';
import _ from "lodash";

const CREATE_DATA = gql`
    mutation CreateAuthorMutation($name: String!, $gender: Boolean!) {
        createAuthor(name: $name, gender: $gender) {
            id
            name
            gender
        }
    }
`;

const Author = (props) => {
    const width = _.get(props, 'width', '');
    const [createAuthor] = useMutation(CREATE_DATA, {
        refetchQueries: [
            props.query, // DocumentNode object parsed with gql
            'getAuthors' // Query name
        ],
    });
    const [dataForm, setDataForm] = useState({
        name: '',
        gender: true
    });

    const save = event => {
        event.preventDefault();
        createAuthor({
                variables: dataForm
            }
        );
        if (_.isFunction(props.onClose)) {
            props.onClose();
        }
    }

    const onChange = (event) => {
        const target = event.target;
        let {name, value} = target;
        if (_.includes(['radio', 'select-one', 'checkbox'], target.type)) {
            if (value === 'true') {
                value = true
            } else if (value === 'false') {
                value = false
            } else if (!_.isNaN(parseInt(value))) {
                value = parseInt(value)
            }
        }

        if (_.isArray(dataForm[name])) {
            const cloneData = dataForm[name];
            if (target.checked) {
                cloneData.push(value)
            } else {
                _.remove(cloneData, function (o) {
                    return o === value;
                })
            }
            value = cloneData;
        }

        setDataForm({
            ...dataForm,
            ...{
                [name]: value
            }
        });
    }

    return (
        <>
            <form className='w-full' onSubmit={save}>
                <table className={`w-1/2 mx-auto ${width}`}>
                    <tbody>
                    <tr>
                        <td>Name</td>
                        <td>
                            <input type="text" name='name' className='border-2 h-10 w-full p-2'
                                   value={dataForm.name}
                                   onChange={onChange}/>
                        </td>
                    </tr>
                    <tr>
                        <td>Gender</td>
                        <td>
                            <div className='flex gap-2 w-full'>
                                <div className='flex-1' key='male'>
                                    <input type="radio" id='male' name='gender' value={true}
                                           onChange={onChange}/>
                                    <label htmlFor='male'>Male</label>
                                </div>
                                <div className='flex-1' key='female'>
                                    <input type="radio" id='female' name='gender' value={false}
                                           onChange={onChange}/>
                                    <label htmlFor='female'>Female</label>
                                </div>
                            </div>
                        </td>
                    </tr>
                    </tbody>
                </table>

                <div className='w-1/2 mt-2'>
                    <button type='submit'
                            className='float-right bg-blue-500 text-white px-3 py-2 rounded-md font-medium'>Save
                    </button>
                </div>
            </form>
        </>
    )
}

export default Author