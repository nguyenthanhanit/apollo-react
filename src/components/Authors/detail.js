import {useQuery, gql} from '@apollo/client';
import _ from "lodash";
import {useParams} from "react-router-dom";

const buildQuery = (id) => {
    return gql`
        {
            getAuthor(id: ${id}) {
                id
                name
                gender
            }
        }
    `;
}

const Author = () => {
    const {id} = useParams();
    const author = _.get(useQuery(buildQuery(id)), 'data.getAuthor', []);

    return <div>
        {!_.isEmpty(author) &&
        <h2>
            Name: {author.name}
            <br/>
            Gender: {author.gender ? 'Male' : 'Female'}
        </h2>
        }
    </div>
}

export default Author