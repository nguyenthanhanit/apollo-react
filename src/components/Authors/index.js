import {useQuery, gql, useMutation} from '@apollo/client';
import _ from "lodash";
import {Link} from "react-router-dom";
import {save} from "../../utils";
import React from "react";

const GET_DATA = gql`
    {
        getAuthors {
            id
            name
            gender
        }
    }
`;

const DELETE_DATA = gql`
    mutation DeleteAuthorMutation($id: ID!) {
        deleteAuthor(id: $id)
    }
`;

const Authors = () => {
    const authors = _.get(useQuery(GET_DATA), 'data.getAuthors', []);
    const [deleteAuthor] = useMutation(DELETE_DATA, {
        refetchQueries: [
            GET_DATA, // DocumentNode object parsed with gql
            'getCategories' // Query name
        ],
    });

    const render = author => {
        return (
            <tr key={author.id}>
                <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">
                    {author.id}
                </td>
                <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">
                    {author.name}
                </td>
                <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">
                    {author.gender ? 'Male' : 'Female'}
                </td>
                <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">
                    <Link to={`author/${author.id}`}
                          className='bg-blue-500 text-white px-3 py-2 rounded-md text-sm font-medium mr-1'>View</Link>
                    <button
                        type="button"
                        className="bg-red-500 text-white px-3 py-2 rounded-md text-sm font-medium"
                        onClick={(event) => save(event, deleteAuthor, {id: author.id})}
                    >
                        Delete
                    </button>
                </td>
            </tr>
        );
    }

    return (
        <div className="flex flex-col">
            <div className="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8">
                <div className="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8">
                    <div className="shadow overflow-hidden border-b border-gray-200 sm:rounded-lg">
                        <table className="min-w-full divide-y divide-gray-200">
                            <thead className="bg-gray-50">
                            <tr>
                                <th scope="col"
                                    className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                                    Id
                                </th>
                                <th scope="col"
                                    className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                                    Name
                                </th>
                                <th scope="col"
                                    className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
                                    Gender
                                </th>
                                <th scope="col" className="relative px-6 py-3">
                                    <span className="sr-only">Action</span>
                                </th>
                            </tr>
                            </thead>
                            <tbody className="bg-white divide-y divide-gray-200">
                            {
                                _.map(authors, render)
                            }
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
        </div>
    );
};

export default Authors;