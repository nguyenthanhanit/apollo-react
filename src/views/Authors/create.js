import StoreForm from "../../containers/StoreForm";

export default function Author() {
    const fields = {
        name: 'String',
        gender: 'Boolean'
    };

    return <StoreForm fields={fields}/>
}