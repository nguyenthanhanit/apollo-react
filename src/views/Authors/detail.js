import {useQuery, gql} from '@apollo/client';
import _ from "lodash";
import {useParams} from "react-router-dom";
import React from "react";

const GET_DATA = gql`
    query Query($id: ID!) {
        getAuthor(id: $id) {
            id
            name
            gender
        }
    }
`;

export default function Author() {
    const {loading, data} = useQuery(GET_DATA, {
        variables: useParams()
    });
    const author = _.get(data, 'getAuthor', []);
    if (loading) {
        return <div>Loading</div>
    }

    return <div>
        {!_.isEmpty(author) &&
        <h2>
            Name: {author.name}
            <br/>
            Gender: {author.gender ? 'Male' : 'Female'}
        </h2>
        }
    </div>
}