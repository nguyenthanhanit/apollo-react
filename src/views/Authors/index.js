import {useQuery, gql, useMutation} from '@apollo/client';
import _ from "lodash";
import React from "react";
import List from "../../components/List";

const GET_DATA = gql`
    {
        getAuthors {
            id
            name
            gender
        }
    }
`;

const DELETE_DATA = gql`
    mutation DeleteAuthorMutation($id: ID!) {
        deleteAuthor(id: $id)
    }
`;

export default function Authors() {
    const authors = _.get(useQuery(GET_DATA), 'data.getAuthors', []);
    const fields = ['name', 'gender'];
    const model = 'Author';
    const [deleteAuthor] = useMutation(DELETE_DATA, {
        refetchQueries: [
            GET_DATA,
            'getAuthors'
        ],
    });

    return <List data={authors} model={model} fields={fields} actionDelete={deleteAuthor}/>
};