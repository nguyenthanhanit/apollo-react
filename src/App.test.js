import {cleanup, render, fireEvent} from '@testing-library/react';
import React from "react";
import Authors from "./views/Authors";
import {ApolloClient, ApolloProvider, createHttpLink, InMemoryCache} from "@apollo/client";
import {Provider} from "react-redux";
import createSagaMiddleware from "redux-saga";
import {applyMiddleware, createStore} from "redux";
import rootReducer from "./reducers/rootReducers";
import rootSaga from "./sagas";
import {BrowserRouter as Router, Route, Switch} from "react-router-dom";
import CreateAuthor from "./views/Authors/create";

const sagaMiddleware = createSagaMiddleware()

// Create store
const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));
sagaMiddleware.run(rootSaga)

const httpLink = createHttpLink({
    uri: 'http://localhost:4000'
});

const client = new ApolloClient({
    link: httpLink,
    cache: new InMemoryCache()
});

afterEach(cleanup);

it('test user form', async () => {
    const {getByText, container} = render(
        <ApolloProvider client={client}>
            <Provider store={store}>
                <Router>
                    <Authors/>
                    <Switch>
                        <Route exact path="/author/create" component={CreateAuthor}/>
                    </Switch>
                </Router>
            </Provider>
        </ApolloProvider>
    );

    expect(getByText(/create/i)).toBeTruthy();
    fireEvent.click(getByText(/Create/));
    expect(getByText(/Save/)).toBeTruthy()
    expect(container).toMatchSnapshot();
});