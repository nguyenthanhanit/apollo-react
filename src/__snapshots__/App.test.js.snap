// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test user form 1`] = `
<div>
  <div>
    <a
      class="bg-blue-500 text-white px-3 py-2 rounded-md text-sm font-medium mr-1"
      href="/Author/Author/create"
    >
      Create
    </a>
    <div
      class="flex flex-col mt-5"
    >
      <div
        class="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8"
      >
        <div
          class="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8"
        >
          <div
            class="shadow overflow-hidden border-b border-gray-200 sm:rounded-lg"
          >
            <table
              class="min-w-full divide-y divide-gray-200"
            >
              <caption>
                Author
              </caption>
              <thead
                class="bg-gray-50"
              >
                <tr>
                  <th
                    class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                    scope="col"
                  >
                    name
                  </th>
                  <th
                    class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                    scope="col"
                  >
                    gender
                  </th>
                  <th
                    class="relative px-6 py-3"
                    scope="col"
                  >
                    <span
                      class="sr-only"
                    >
                      Action
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                class="bg-white divide-y divide-gray-200"
              />
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
  <form
    class="w-full"
  >
    <table
      class="w-1/2 mx-auto"
    >
      <tbody>
        <tr>
          <td
            class="uppercase"
          >
            name
          </td>
          <td>
            <input
              class="border-2 h-10 w-full p-2"
              name="name"
              type="text"
              value=""
            />
          </td>
        </tr>
        <tr>
          <td
            class="uppercase"
          >
            gender
          </td>
          <td>
            <div
              class="flex gap-2 w-full"
            >
              <div
                class="flex-1"
              >
                <input
                  id="male"
                  name="gender"
                  type="radio"
                  value="true"
                />
                <label
                  for="male"
                >
                  Male
                </label>
              </div>
              <div
                class="flex-1"
              >
                <input
                  id="female"
                  name="gender"
                  type="radio"
                  value="false"
                />
                <label
                  for="female"
                >
                  Female
                </label>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="w-1/2 mt-2"
    >
      <button
        class="float-right bg-blue-500 text-white px-3 py-2 rounded-md font-medium"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
